cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +678,78 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-// s.mu must be locked when calling handleProcessExit
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
-			}
+// handleInitExit processes container init process exits.
+// This is handled separately from non-init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - for a given container, the init process exit MUST be the last exit published
+// This is achieved by:
+// - killing all running container processes (if the container has a shared pid)
+// namespace, otherwise all other processes have been reaped already.
+// - waiting for the container's running exec counter to reach 0, or timeout.
+// - removing all remaining processes from s.running so that we s.processExits
+// doesn't publish their exits.
+// - publishing the received exec exits, and then publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	// kill all running container processes
+	if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
+		if err := p.KillAll(s.context); err != nil {
+			log.G(s.context).WithError(err).WithField("id", p.ID()).
+				Error("failed to kill init's children")
 		}
 	}
 
+	s.lifecycleMu.Lock()
+	defer s.lifecycleMu.Unlock()
+	numRunningExecs := s.runningExecs[c]
+	if numRunningExecs == 0 {
+		s.handleProcessExit(e, c, p)
+		return
+	}
+
+	events := make(chan int, numRunningExecs)
+	s.execCountSubscribers[c] = events
+
+	go func() {
+		defer func() {
+			s.lifecycleMu.Lock()
+			defer s.lifecycleMu.Unlock()
+			delete(s.execCountSubscribers, c)
+		}()
+
+		// wait for running processes to exit
+		for runningExecs := range events {
+			if runningExecs == 0 {
+				break
+			}
+		}
+
+		// wipe all remaining processes from `s.running` so that late-reaped
+		// exec exits won't be published after the init exit
+		for _, execProcess := range c.ExecdProcesses() {
+			pid := execProcess.Pid()
+			if _, running := s.running[execProcess.Pid()]; running {
```

Comment:
Access to `s.running` needs to be guarded by `s.lifecycleMu`, right?  Looks like that's missing here.